src/compiler/transformers/classFields.ts

```diff
@@ -439,23 +567,39 @@ namespace ts {
         function visitClassLike(node: ClassLikeDeclaration) {
             const savedPendingExpressions = pendingExpressions;
             pendingExpressions = undefined;
-            if (shouldTransformPrivateFields) {
+            if (shouldTransformPrivateElements) {
                 startPrivateIdentifierEnvironment();
+
+                const name = getNameOfDeclaration(node);
+                if (name && isIdentifier(name)) {
+                    getPrivateIdentifierEnvironment().className = name.escapedText as string;
+                }
+
+                if (some(getPrivateInstanceMethods(node))) {
+                    getPrivateIdentifierEnvironment().weakSetName = createHoistedVariableForClass("instances");
+                }
             }
 
             const result = isClassDeclaration(node) ?
                 visitClassDeclaration(node) :
                 visitClassExpression(node);
 
-            if (shouldTransformPrivateFields) {
+            if (shouldTransformPrivateElements) {
                 endPrivateIdentifierEnvironment();
             }
             pendingExpressions = savedPendingExpressions;
             return result;
         }
 
         function doesClassElementNeedTransform(node: ClassElement) {
-            return isPropertyDeclaration(node) || (shouldTransformPrivateFields && node.name && isPrivateIdentifier(node.name));
+            return isPropertyDeclaration(node) || (shouldTransformPrivateElements && node.name && isPrivateIdentifier(node.name));
+        }
+
+        function getPrivateInstanceMethods(node: ClassLikeDeclaration) {
+            return filter(

```

Comment:
After reviewing utilities.ts, you might want to consider just creating an `isNonStaticMethodOrAccessorWithPrivateName` function rather than all of the extra functions and closures needed to evaluate this. This implementation generates:
- Two function definitions that are only used from here (`getMethods` and `getAccessors`)
- Two helper functions that are only used from the two functions above (`isStaticMethodDeclaration` and `isStaticAccessorDeclaration`)
- A closure a piece in each of `getMethods` and `getAccessors` to pass the `isStatic` flag around.
- An arrow function for the final filter here.

It seems like this complexity could be greatly reduced.